File: smu/geometry/smu_molecule_test.jl
"""Testing the SmuMolecule.
The 1 based array indexing in Julia makes this somewhat discordant with the phython
implementation.
"""

module TestSmuMolecule

using ProtoBuf
using Test

using SmuMolecules
using SmuUtilities

using dataset_pb2

export all_tests

"""Make a BondTopology consisting of `n` disconnected Carbon atoms.
"""
function carbon_atoms(n::T)::BondTopology where {T<:Integer}
  result = BondTopology()
  for i in 1:n
    add_atom_by_atomic_number!(result, 6)
  end
  return result
end

"""The simplest multi-atom molecule, CC"""
function test_ethane()
  matching_parameters = MatchingParameters()
  matching_parameters.must_match_all_bonds = false

  ethane = carbon_atoms(2)
  add_bond!(0, 1, 1, ethane)
  scores = [0.1, 1.1, 2.1, 3.1]
  bonds_to_scores = Dict{Tuple{Int32, Int32}, Vector{Float32}}((1, 2) => scores)
  mol = SmuMolecule(ethane, bonds_to_scores, matching_parameters)
  state = generate_search_state(mol)
  @test length(state) == 1
  @test state == [[1, 2, 3, 4]]

  for s in Iterators.product(state...)
    res = place_bonds!(s, mol)
    @test res !== nothing
    # self.assertAlmostEqual(res.score, scores[i])
  end
  true
end

function test_ethane_all_btypes()
  matching_parameters = MatchingParameters()
  matching_parameters.must_match_all_bonds = false

  test_cases = [
    [0, dataset_pb2.BondTopology_BondType.BOND_UNDEFINED],
    [1, dataset_pb2.BondTopology_BondType.BOND_SINGLE],
    [2, dataset_pb2.BondTopology_BondType.BOND_DOUBLE],
    [3, dataset_pb2.BondTopology_BondType.BOND_TRIPLE],
  ]

  for t in test_cases
    numeric_btype = t[1]
    smu_btype = t[2]
    cc = carbon_atoms(2)
    bonds_to_scores = Dict{Tuple{Int32,Int32}, Vector{Float32}}((1, 2) => zeros(Float32, 4))
    bonds_to_scores[(1, 2)][numeric_btype + 1] = 1.0
    mol = SmuMolecule(cc, bonds_to_scores, matching_parameters)
    state = generate_search_state(mol)
    for s in Iterators.product(state...)
      res = place_bonds!(s, mol)
      @test res !== nothing
      if numeric_btype == 0
        @test !hasproperty(res, :bonds)
      else
        @test length(res.bonds) == 1
        @test res.bonds[1].bond_type == smu_btype
      end
    end
  end
  true
end

"""Checking to see if a BondTopology has the right number of bonds is
complicated by the fact that if there are no bonds present, the :bonds
attribute will be missing.
Args:
  bond_topology:
  expected_bonds: the number of bonds expected to be in `bond_topology`.
Returns:
  true if the number of bonds present matches `expected_bonds`.
"""
function bond_count_matches(bond_topology::BondTopology, expected_bonds)::Bool
  if !hasproperty(bond_topology, :bonds)
    return expected_bonds == 0
  end

  return length(bond_topology.bonds) == expected_bonds
end

struct A1A2BtypeScore
  btype1::Int32
  btype2::Int32
  expected_bonds::Int32
  expected_score::Union{Float32, Nothing}
end

function test_propane_all()
  matching_parameters = MatchingParameters()
  matching_parameters.must_match_all_bonds = false

  test_cases = [
    A1A2BtypeScore(0, 0, 0, 2.0),
    A1A2BtypeScore(0, 1, 1, 2.0),
    A1A2BtypeScore(0, 2, 1, 2.0),
    A1A2BtypeScore(0, 3, 1, 2.0),
    A1A2BtypeScore(1, 1, 2, 2.0),
    A1A2BtypeScore(1, 2, 2, 2.0),
    A1A2BtypeScore(1, 3, 2, 2.0),
    A1A2BtypeScore(2, 2, 2, 2.0),
    A1A2BtypeScore(2, 3, 0, nothing),
    A1A2BtypeScore(3, 3, 0, nothing)
  ]
  for t in test_cases
    btype1 = t.btype1
    btype2 = t.btype2
    expected_bonds = t.expected_bonds
    expected_score = t.expected_score
    cc = carbon_atoms(3)
    bonds_to_scores = Dict{Tuple{Int32,Int32}, Vector{Float32}}((1, 2) => zeros(Float32, 4),
                       (2, 3) => zeros(Float32, 4))
    bonds_to_scores[(1, 2)][btype1 + 1] = 1.0
    bonds_to_scores[(2, 3)][btype2 + 1] = 1.0
    mol = SmuMolecule(cc, bonds_to_scores, matching_parameters)
    state = generate_search_state(mol)
    for s in Iterators.product(state...)
      res = place_bonds!(s, mol)
      if expected_score !== nothing
        @test res !== nothing
        @test bond_count_matches(res, expected_bonds)
        @test isapprox(res.score, expected_score)
        if btype1 == 0
          if btype2 > 0
            @test res.bonds[1].bond_type == btype2
          end
        else
          @test Set((btype1, btype2)) == Set((res.bonds[1].bond_type, res.bonds[2].bond_type))
        end
      else
        @test res === nothing
      end
    end
  end
  true
end

function test_operators()
  matching_parameters = MatchingParameters()
  matching_parameters.must_match_all_bonds = false

  cc = carbon_atoms(3)
  add_bond!(0, 1, 1, cc)
  add_bond!(1, 2, 1, cc)
  bonds_to_scores = Dict{Tuple{Int32,Int32}, Vector{Float32}}((1, 2) => zeros(Float32, 4),
                                            (2, 3) => zeros(Float32, 4))
  scores = [1.0, 3.0]
  bonds_to_scores[(1, 2)][1] = scores[1]
  bonds_to_scores[(2, 3)][1] = scores[2]
  mol = SmuMolecule(cc, bonds_to_scores, matching_parameters)
  set_initial_score_and_incrementer!(1.0, Base.:*, mol)
  state = generate_search_state(mol)
  for s in Iterators.product(state...)
    res = place_bonds!(s, mol)
    @test isapprox(res.score, prod(scores))
  end
  true
end


function all_tests()
  @test test_ethane()
  @test test_ethane_all_btypes()
  @test test_propane_all()
  @test test_operators()
end



end  # module TestSmuMolecule

File: smu/geometry/smu_molecule_test.py
# coding=utf-8
# Copyright 2022 The Google Research Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import itertools
import operator

from absl.testing import absltest
import numpy as np
from parameterized import parameterized

from google.protobuf import text_format
from smu import dataset_pb2
from smu.geometry import smu_molecule


class TestSmuMolecule(absltest.TestCase):
  """Test the SmuMolecule class."""

  @absltest.skip("This test is duplciative of test_ethane_all")
  def test_ethane(self):
    """The simplest molecule, CC."""
    #     bond_topology = text_format.Parse(
    #         """
    #       atoms: ATOM_C
    #       atoms: ATOM_C
    #       bonds: {
    #         atom_a: 0,
    #         atom_b: 1,
    #         bond_type: BOND_SINGLE
    #       }
    # """, dataset_pb2.BondTopology())
    cc = text_format.Parse("""
      atoms: ATOM_C
      atoms: ATOM_C
""", dataset_pb2.BondTopology())
    scores = np.array([0.1, 1.1, 2.1, 3.1], dtype=np.float32)
    bonds_to_scores = {(0, 1): scores}
    matching_parameters = smu_molecule.MatchingParameters()
    matching_parameters.must_match_all_bonds = False
    mol = smu_molecule.SmuMolecule(cc, bonds_to_scores, matching_parameters)
    state = mol.generate_search_state()
    self.assertLen(state, 1)
    self.assertEqual(state, [[0, 1, 2, 3]])

    for i, s in enumerate(itertools.product(*state)):
      res = mol.place_bonds(s)
      self.assertIsNotNone(res)
      self.assertAlmostEqual(res.score, scores[i])

  @parameterized.expand([
      [0, dataset_pb2.BondTopology.BOND_UNDEFINED],
      [1, dataset_pb2.BondTopology.BOND_SINGLE],
      [2, dataset_pb2.BondTopology.BOND_DOUBLE],
      [3, dataset_pb2.BondTopology.BOND_TRIPLE],
  ])
  def test_ethane_all(self, btype, expected_bond):
    cc = text_format.Parse("""
      atoms: ATOM_C
      atoms: ATOM_C
""", dataset_pb2.BondTopology())
    bonds_to_scores = {(0, 1): np.zeros(4, dtype=np.float32)}
    bonds_to_scores[(0, 1)][btype] = 1.0
    matching_parameters = smu_molecule.MatchingParameters()
    matching_parameters.must_match_all_bonds = False
    mol = smu_molecule.SmuMolecule(cc, bonds_to_scores, matching_parameters)
    state = mol.generate_search_state()
    for s in itertools.product(*state):
      res = mol.place_bonds(s, matching_parameters)
      if btype == 0:
        self.assertIsNone(res)
      else:
        self.assertIsNotNone(res)
        self.assertLen(res.bonds, 1)
        self.assertEqual(res.bonds[0].bond_type, expected_bond)

  @parameterized.expand([
      [0, 0, 0, None],
      [0, 1, 1, None],
      [0, 2, 1, None],
      [0, 3, 1, None],
      [1, 1, 2, 1.0],
      [1, 2, 2, 1.0],
      [1, 3, 2, 1.0],
      [2, 2, 2, 1.0],
      [2, 3, 0, None],
      [3, 3, 0, None],
  ])
  def test_propane_all(self, btype1, btype2, expected_bonds, expected_score):
    cc = text_format.Parse(
        """
      atoms: ATOM_C
      atoms: ATOM_C
      atoms: ATOM_C
""", dataset_pb2.BondTopology())
    #   print(f"Generating bonds {btype1} and {btype2}")
    bonds_to_scores = {
        (0, 1): np.zeros(4, dtype=np.float32),
        (1, 2): np.zeros(4, dtype=np.float32)
    }
    bonds_to_scores[(0, 1)][btype1] = 1.0
    bonds_to_scores[(1, 2)][btype2] = 1.0
    matching_parameters = smu_molecule.MatchingParameters()
    matching_parameters.must_match_all_bonds = False
    mol = smu_molecule.SmuMolecule(cc, bonds_to_scores, matching_parameters)
    state = mol.generate_search_state()
    for s in itertools.product(*state):
      res = mol.place_bonds(s, matching_parameters)
      if expected_score is not None:
        self.assertIsNotNone(res)
        self.assertLen(res.bonds, expected_bonds)
        self.assertAlmostEqual(res.score, expected_score)
        if btype1 == 0:
          if btype2 > 0:
            self.assertEqual(res.bonds[0].bond_type, btype2)
        else:
          self.assertEqual(res.bonds[0].bond_type, btype1)
          self.assertEqual(res.bonds[1].bond_type, btype2)
      else:
        self.assertIsNone(res)

  def test_operators(self):
    cc = text_format.Parse(
        """
      atoms: ATOM_C
      atoms: ATOM_C
      atoms: ATOM_C
""", dataset_pb2.BondTopology())
    #   print(f"Generating bonds {btype1} and {btype2}")
    bonds_to_scores = {
        (0, 1): np.zeros(4, dtype=np.float32),
        (1, 2): np.zeros(4, dtype=np.float32)
    }
    scores = np.array([1.0, 3.0], dtype=np.float32)
    bonds_to_scores[(0, 1)][1] = scores[0]
    bonds_to_scores[(1, 2)][1] = scores[1]
    matching_parameters = smu_molecule.MatchingParameters()
    matching_parameters.must_match_all_bonds = False
    mol = smu_molecule.SmuMolecule(cc, bonds_to_scores, matching_parameters)
    mol.set_initial_score_and_incrementer(1.0, operator.mul)
    state = mol.generate_search_state()
    for s in itertools.product(*state):
      res = mol.place_bonds(s, matching_parameters)
      self.assertAlmostEqual(res.score, np.product(scores))


if __name__ == "__main__":
  absltest.main()
